File: server/src/main/java/neuroevolution/ann/NeuralNetwork.java
package neuroevolution.ann;

import neuroevolution.exception.InvalidDimensionsException;
import neuroevolution.math.Vector;

import java.util.List;

public class NeuralNetwork {

	private final List<NeuralNetworkLayer> layers;

	public NeuralNetwork(final List<NeuralNetworkLayer> layers) {
		for (int i = 1; i < layers.size(); ++i) {
			if (layers.get(i - 1).outputSize() != layers.get(i).inputSize()) {
				throw new InvalidDimensionsException("Dimensions of adjacent layers does not match.");
			}
		}
		this.layers = layers;
	}

	public Vector evaluate(final Vector input) {
		return layers.stream().reduce(input, (vector, layer) -> layer.evaluate(vector), (accumulator, value) -> value);
	}

}

File: webapp/src/app/evolution/logic/neural-network/neural-network.ts
import {NeuralNetworkLayer} from './neural-network-layer';
import {NeuralNetworkSettings} from './settings/neural-network-settings';
import {NeuralNetworkLayerSettings} from './settings/neural-network-layer-settings';
import {Matrix} from './matrix/matrix';
import {Vector} from './vector/vector';

export class NeuralNetwork {

  readonly layers: readonly NeuralNetworkLayer[];

  constructor(settings: NeuralNetworkSettings, elements: Iterator<number>) {
    this.layers = settings.layers.map((layer: NeuralNetworkLayerSettings, index: number) => new NeuralNetworkLayer(
      Matrix.fromProvider(
        layer.numberOfNeurons,
        index === 0 ? settings.inputSize : settings.layers[index - 1].numberOfNeurons,
        (): number => elements.next().value
      ),
      Vector.fromProvider(
        layer.numberOfNeurons,
        (): number => elements.next().value
      ),
      layer.activationFunction
    ));
  }

  evaluate(input: Vector): Vector {
    return this.layers.reduce((accumulator: Vector, layer: NeuralNetworkLayer): Vector => layer.evaluate(accumulator), input);
  }

}
